metals/src/main/scala/scala/meta/internal/metals/PostCompileCache.scala

```diff
@@ -0,0 +1,60 @@
+package scala.meta.internal.metals
+
+import java.util.Collections.singletonList
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.atomic.AtomicBoolean
+import ch.epfl.scala.{bsp4j => b}
+import scala.meta.internal.metals.MetalsEnrichments._
+
+final class PostCompileCache(buildServer: () => Option[BuildServerConnection]) {
+  private val initialized = new AtomicBoolean(false)
+  val mainClasses = new ConcurrentHashMap[String, b.ScalaMainClass]()
```

Comment:
make it per buildTarget
e.g. 
```scala
private class EntryPoints{
  val mainClasses = new ConcurrentHashMap[String, b.ScalaMainClass]()
}

val entryPoints = new ConcurrentHashMap[BuildTargetIdentifier, EntryPoints]()
```

Reply:
What are the keys in this map? We BTW normally use `TrieMap` by default for indexes and only `ConcurrentHashMap` if there's a specific need for it.

Reply:
there are three threads using those structures:
1. one actually fetching main classes
2. one listening for bsp messages
3. one asking for main classes to create code lenses